File: BlockBattleBot/Parser.cs
using System;
using System.Linq;

namespace BlockBattleBot
{
    public class Parser
    {
        public Game Game { get; private set; }

        public Brain Brain { get; private set; }

        public Parser(Game game, Brain brain)
        {
            Game = game;
            Brain = brain;
        }

        public object Parse(string line)
        {
            string response = null;

            if (!string.IsNullOrWhiteSpace(line))
            {
                string[] segments = line.Split(' ');

                if (segments.Length < 3)
                {
                    throw new Exception(string.Format("Expected 3 segments, got {0}.", segments.Length));
                }

                switch (segments[0])
                {
                    case "settings":
                        switch (segments[1])
                        {
                            case "timebank":
                                Game.Settings.MaxTimeBank = Convert.ToInt32(segments[2]);
                                break;
                            case "time_per_move":
                                Game.Settings.TimePerMove = Convert.ToInt32(segments[2]);
                                break;
                            case "player_names":
                                Game.Players.Clear();
                                string[] names = segments[2].Split(',');
                                foreach (var name in names)
                                {
                                    Game.Players.Add(name, new Player(0, 0, new Field(Game.Settings.FieldWidth, Game.Settings.FieldHeight)));
                                }
                                break;
                            case "your_bot":
                                Game.PlayerName = segments[2];
                                break;
                            case "field_width":
                                Game.Settings.FieldWidth = Convert.ToInt32(segments[2]);
                                break;
                            case "field_height":
                                Game.Settings.FieldHeight = Convert.ToInt32(segments[2]);
                                break;
                        }
                        break;

                    case "update":
                        if (segments[1] == "game")
                        {
                            switch (segments[2])
                            {
                                case "round":
                                    Game.Round.Number = Convert.ToInt32(segments[3]);
                                    break;
                                case "this_piece_type":
                                    Game.Round.Piece = (PieceType)Enum.Parse(typeof(PieceType), segments[3]);
                                    break;
                                case "next_piece_type":
                                    Game.Round.NextPiece = (PieceType)Enum.Parse(typeof(PieceType), segments[3]);
                                    break;
                                case "this_piece_position":
                                    int[] points = segments[3].Split(',').Select(point => Convert.ToInt32(point)).ToArray();
                                    Game.Round.PiecePosition = new Position(points[0], points[1]);
                                    break;
                            }
                        }
                        else
                        {
                            switch (segments[2])
                            {
                                case "row_points":
                                    Game.Players[segments[1]].Points = Convert.ToInt32(segments[3]);
                                    break;
                                case "combo":
                                    Game.Players[segments[1]].Combo = Convert.ToInt32(segments[3]);
                                    break;
                                case "field":
                                    Game.Players[segments[1]].Field.Reset(Game.Settings.FieldWidth, Game.Settings.FieldHeight);

                                    string[] rows = segments[3].Trim(';').Split(';');

                                    for (int y = 0; y < rows.Length; y++)
                                    {
                                        string[] columns = rows[y].Split(',');
                                        for (int x = 0; x < columns.Length; x++)
                                        {
                                            CellStatus status = (CellStatus)Enum.Parse(typeof(CellStatus), columns[x]);
                                            Game.Players[segments[1]].Field.SetCell(x, y, status);
                                        }
                                    }
                                    break;
                            }
                        }
                        break;

                    case "action":
                        if (segments[1] == "moves")
                        {
                            response = string.Join(",", Brain.Moves(Convert.ToInt32(segments[2])).Select(m => m.ToString().ToLower()));
                        }
                        break;

                    default:
                        throw new Exception(string.Format("Unknown command: {0}.", segments[0]));
                }
            }

            return response;
        }
    }
}

File: src/Parser.ts
import { Brain } from "./Brain";
import { CellStatus } from "./CellStatus";
import { Game } from "./Game";
import { Move } from "./Move";
import { PieceType, parse as parsePieceType } from "./PieceType";
import { Position } from "./Position";

export class Parser {
	private _game: Game;
	private _brain: Brain;
	
	constructor(game: Game, brain: Brain) {
		this._game = game;
		this._brain = brain;
	}
	
	public parse(line: string): Array<Move> {
		let response: any = null;
		
		line = line.trim();
		
		if (line.length > 0) {
			let segments: Array<string> = line.split(" ");
			
			switch (segments[0]) {
				case "settings":
					if (segments.length < 3) {
						throw new Error("Settings command requires 2 arguments.");
					}
				
					this.handleSettings(segments[1], segments[2]);
					break;
				case "update":
					if (segments.length < 4) {
						throw new Error("Update command requires 3 arguments.");
					}
				
					this.handleUpdate(segments[1], segments[2], segments[3]);
					break;
				case "action":
					if (segments.length < 2) {
						throw new Error("Action command requires 1 argument.");
					}
				
					if (segments[1] == "moves") {
						response = this._brain.moves(parseInt(segments[2]));
					}
					break;
			}
		}
		
		return response;
	}
	
	private handleSettings(setting: string, value: string): void {
		switch (setting) {
            case "timebank":
				this._game.settings.maxTimeBank = parseInt(value);
                break;
            case "time_per_move":
				this._game.settings.timePerMove = parseInt(value);
                break;
            case "player_names":
				let playerNames: string[] = value.split(",");
				playerNames.forEach((name: string) => {
					this._game.addPlayer(name);
				});
                break;
            case "your_bot":
				this._game.playerName = value;
                break;
            case "field_width":
				this._game.settings.fieldWidth = parseInt(value);
                break;
            case "field_height":
				this._game.settings.fieldHeight = parseInt(value);
                break;
        }
	}
	
	private handleUpdate(entity: string, field: string, value: string): void {
		if (entity == "game") {
			switch (field) {
                case "round":
					this._game.round.number = parseInt(value);
                    break;
                case "this_piece_type":
					this._game.round.piece = parsePieceType(value);
                    break;
                case "next_piece_type":
					this._game.round.nextPiece = parsePieceType(value);
                    break;
                case "this_piece_position":
					let points = value.split(",");
					this._game.round.startPosition = new Position(parseInt(points[0]), parseInt(points[1]));
                    break;
            }
		} else {
			switch (field) {
				case "row_points":
					this._game.players[entity].points = parseInt(value);
                    break;
                case "combo":
					this._game.players[entity].combo = parseInt(value);
                    break;
                case "field":
					this._game.players[entity].field.reset(this._game.settings.fieldWidth, this._game.settings.fieldHeight);
				
					let rows = value.replace(/;$/, "").split(";");
					
					for (let y = 0; y < rows.length; y++) {
						let columns= rows[y].split(",");
						
						for (let x = 0; x < columns.length; x++) {
							let cellStatus = <CellStatus>parseInt(columns[x])
							this._game.players[entity].field.setCell(x, y, cellStatus);
						}
					}
                    break;
			}
		}
	}
}